p2p/src/secret_connection.rs

```diff
@@ -0,0 +1,420 @@
+//! `SecretConnection`: Transport layer encryption for Tendermint P2P connections.
+
+mod amino_types;
+mod kdf;
+mod nonce;
+mod protocol;
+mod public_key;
+
+pub use self::{kdf::Kdf, nonce::Nonce, protocol::Version, public_key::PublicKey};
+use crate::error::Error;
+use chacha20poly1305::{
+    aead::{generic_array::GenericArray, AeadInPlace, NewAead},
+    ChaCha20Poly1305,
+};
+use ed25519_dalek::{self as ed25519, Signer, Verifier};
+use eyre::{Result, WrapErr};
+use merlin::Transcript;
+use rand_core::OsRng;
+use std::{
+    cmp,
+    convert::{TryFrom, TryInto},
+    io::{self, Read, Write},
+    marker::{Send, Sync},
+    slice,
+};
+use subtle::ConstantTimeEq;
+use tendermint_proto as proto;
+use x25519_dalek::{EphemeralSecret, PublicKey as EphemeralPublic};
+
+/// Size of the MAC tag
+pub const TAG_SIZE: usize = 16;
+
+/// Maximum size of a message
+pub const DATA_MAX_SIZE: usize = 1024;
+
+/// 4 + 1024 == 1028 total frame size
+const DATA_LEN_SIZE: usize = 4;
+const TOTAL_FRAME_SIZE: usize = DATA_MAX_SIZE + DATA_LEN_SIZE;
+
+/// Encrypted connection between peers in a Tendermint network
+pub struct SecretConnection<IoHandler: Read + Write + Send + Sync> {
+    io_handler: IoHandler,
+    protocol_version: Version,
+    recv_nonce: Nonce,
+    send_nonce: Nonce,
+    recv_cipher: ChaCha20Poly1305,
+    send_cipher: ChaCha20Poly1305,
+    remote_pubkey: Option<PublicKey>,
+    recv_buffer: Vec<u8>,
+}
+
+impl<IoHandler: Read + Write + Send + Sync> SecretConnection<IoHandler> {
+    /// Returns authenticated remote pubkey
+    pub fn remote_pubkey(&self) -> PublicKey {
+        self.remote_pubkey.expect("remote_pubkey uninitialized")
+    }
+
+    /// Performs handshake and returns a new authenticated SecretConnection.
+    pub fn new(
+        mut io_handler: IoHandler,
+        local_privkey: &ed25519::Keypair,
+        protocol_version: Version,
+    ) -> Result<SecretConnection<IoHandler>> {
+        let local_pubkey = PublicKey::from(local_privkey);
+
+        // Generate ephemeral keys for perfect forward secrecy.
+        let (local_eph_pubkey, local_eph_privkey) = gen_eph_keys();
+
+        // Write local ephemeral pubkey and receive one too.
+        // NOTE: every 32-byte string is accepted as a Curve25519 public key
+        // (see DJB's Curve25519 paper: http://cr.yp.to/ecdh/curve25519-20060209.pdf)
```

Comment:
This note is probably superfluous as the code in `protocol.rs` rejects low order points